tls/s2n_record_write.c

```diff
@@ -250,63 +305,18 @@ int s2n_record_write(struct s2n_connection *conn, uint8_t content_type, struct s
     /* Skip the header */
     GUARD(s2n_stuffer_skip_write(&conn->out, S2N_TLS_RECORD_HEADER_LENGTH));
 
-    uint16_t encrypted_length = data_bytes_to_take + mac_digest_size;
-    switch (cipher_suite->record_alg->cipher->type) {
-	case S2N_AEAD:
-		GUARD(s2n_stuffer_skip_write(&conn->out, cipher_suite->record_alg->cipher->io.aead.record_iv_size));
-		encrypted_length += cipher_suite->record_alg->cipher->io.aead.tag_size;
-		break;
-	case S2N_CBC:
-		if (conn->actual_protocol_version > S2N_TLS10) {
-			/* Leave the IV alone and unencrypted */
-			GUARD(s2n_stuffer_skip_write(&conn->out, iv.size));
-		}
-		/* Encrypt the padding and the padding length byte too */
-		encrypted_length += padding + 1;
-		break;
-	case S2N_COMPOSITE:
-		/* Composite CBC expects a pointer starting at explicit IV: [Explicit IV | fragment | MAC | padding | padding len ]
-		 * extra will account for the explicit IV len(if applicable), MAC digest len, padding len + padding byte.
-		 */
-		encrypted_length += extra;
-		break;
-	default:
-		break;
-    }
+    uint16_t encrypted_length = s2n_record_skip_write_encrypted_length(conn, cipher_suite, iv, padding, extra, data_bytes_to_take + mac_digest_size);
 
     /* Do the encryption */
     struct s2n_blob en;
     en.size = encrypted_length;
     en.data = s2n_stuffer_raw_write(&conn->out, en.size);
     notnull_check(en.data);
 
-    switch (cipher_suite->record_alg->cipher->type) {
-    case S2N_STREAM:
-        GUARD(cipher_suite->record_alg->cipher->io.stream.encrypt(session_key, &en, &en));
-        break;
-    case S2N_CBC:
-        GUARD(cipher_suite->record_alg->cipher->io.cbc.encrypt(session_key, &iv, &en, &en));
-
-        /* Copy the last encrypted block to be the next IV */
-        if (conn->actual_protocol_version < S2N_TLS11) {
-            gte_check(en.size, block_size);
-            memcpy_check(implicit_iv, en.data + en.size - block_size, block_size);
-        }
-        break;
-    case S2N_AEAD:
+    if (cipher_suite->record_alg->cipher->type == S2N_AEAD) {
```

Comment:
What was the rationale for separating out the S2N_AEAD case? To me it seems natural that you would handle each case in s2n_record_write_encryption